src/weapon.c

```diff
@@ -468,17 +468,21 @@ static struct obj *oselect(struct monst *, int);
     } while (0)
 
 static struct obj *
-oselect(struct monst *mtmp, int x)
+oselect(struct monst *mtmp, int type)
 {
     struct obj *otmp;
 
     for (otmp = mtmp->minvent; otmp; otmp = otmp->nobj) {
-        if (otmp->otyp == x
-            /* never select non-cockatrice corpses */
-            && !((x == CORPSE || x == EGG)
-                 && !touch_petrifies(&mons[otmp->corpsenm]))
-            && (!otmp->oartifact || touch_artifact(otmp, mtmp)))
-            return otmp;
+        if (otmp->otyp != type)
+            continue;
+        /* never select non-cockatrice corpses */
+        if ((type == CORPSE || type == EGG)
+            && (otmp->corpsenm == NON_PM
+                || !touch_petrifies(&mons[otmp->corpsenm])))
+            continue;
+        if (otmp->oartifact || !touch_artifact(otmp, mtmp))
```

Comment:
This should be

```c
        if (otmp->oartifact && !touch_artifact(otmp, mtmp))
```

(Actually, maybe it should really be `if (!can_touch_safely(mtmp, otmp))` since that handles other cases like silver-haters vs silver weapons, in addition to `touch_artifact`... Not sure if there's a reason that isn't used or not.)

Reply:
Yup, can_touch_safely seems to contain the artifact check so included that one.